### PR TITLE
ENH: Support Table.asof_join

### LINKED ISSUE #1118: merge_asof for Ibis Tables

### PR DESCRIPTION
**Pull request by toryhaavik**
closes #1118 with a Pandas implementation

### CODE REVIEW

**Comment by toryhaavik** [comment]
I'm realizing merge_asof may not be available in the versions of pandas for 27 and 34.

**Comment by jreback** [comment]
it’s available in 2.7

3.4 is not supported in pandas 0.21 but not out yet

it should be available in 0.19.2 or after

**Comment by toryhaavik** [comment]
Odd.  The 2.7 build fails with `AttributeError: 'module' object has no attribute 'merge_asof'` and uses `pandas==0.18.1`.  The [docs](http://pandas.pydata.org/pandas-docs/version/0.18.1/search.html?q=merge_asof&check_keywords=yes&area=default) don't have any reference to merge_asof in 0.18.1

the 3.4 build fails with `TypeError: merge_asof() got an unexpected keyword argument 'left_by'` and uses `pandas==0.19.0`.  Seems `left_by` was added after 0.19.0

File: ibis/pandas/tests/test_operations.py [hunk 1]
```diff
@@ -50,6 +50,32 @@ def test_join(how, left, right, df1, df2):
     tm.assert_frame_equal(result[expected.columns], expected)
 
 
+def test_asof_join(time_left, time_right, time_df1, time_df2):
```
Comment (cpcloud) [inline ibis/pandas/tests/test_operations.py]: I would just skip these tests on the versions of pandas we test that don't have `merge_asof`.

Comment (toryhaavik) [reply ibis/pandas/tests/test_operations.py]: something like this?

    merge_asof_minversion = pytest.mark.skipif(
        pd.__version__ < '0.19.2',
        reason="at least pandas-0.19.2 required for merge_asof")

    @merge_asof_minversion
    def test_asof_join...

File: ibis/pandas/execution.py [hunk 2]
```diff
@@ -591,23 +591,51 @@ def execute_materialized_join(op, left, right, **kwargs):
...
+    }
+    left_by, right_by = _extract_predicate_names(op.by_predicates)
+    _validate_columns(
+        overlapping_columns, left_on, right_on, left_by, right_by)
+
+    if left_by and right_by:
+        merge_asof_args['left_by'] = left_by
+        merge_asof_args['right_by'] = right_by
+
+    return pd.merge_asof(**merge_asof_args)
```
[the remark this answers is not in the document]
Comment (toryhaavik) [reply ibis/pandas/execution.py]: I wasn't sure what the behavior would be if you pass `left_by=[]` and `right_by=[]`.  I imagine it's a no-op, but I was trying to be safe.  Didn't think about the debug trade off.

[the remark this answers is not in the document]
Comment (toryhaavik) [reply ibis/pandas/execution.py]: Yeah, a quick test tells me that passing empty lists is not the same as passing None.  So the choice is either the dict or something like this that gets passed in as args.  I'm open to either.

    left_by = left_by if left_by else None
    right_by = right_by if right_by else None

[the remark this answers is not in the document]
Comment (toryhaavik) [reply ibis/pandas/execution.py]: Now that I look at it without the dict, it looks much better.  I'll push up the change.

[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/pandas/execution.py]: probably want to do something like:

```python
pd.merge_asof(..., left_by=left_by or None, right_by=right_by or None)
```

**Comment by cpcloud** [comment]
LGTM, merging.

**Comment by cpcloud** [comment]
Thanks @toryhaavik!

**Pull request closed without merging**